### PR TITLE
feat(source): support parsing nested record/array in avro parser and fix bug when load schema from s3

### PR DESCRIPTION
**Pull request by waruto210**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- support parsing nested record/array in avro parser
- enhance the test schema `complex-schema`
- fix a bug when load schema from s3

## Checklist
~- [ ] I have written necessary rustdoc comments~
~- [ ] I have added necessary unit tests and integration tests~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation
Make avro data source support nested record/list, and can load avro schema from s3 storage.

### Types of user-facing changes
- Connector (sources & sinks)

### Release note

Create a materilized source with row format avro and schema on s3:
```SQL
create materialized source cs with ( 
    connector = 'kafka', 
    topic = 'avro_bin_c', 
    properties.bootstrap.server = '127.0.0.1:29092', 
    scan.startup.mode = 'earliest', region = 'us-east-1', 
    endpoint_url='http://127.0.0.1:9301')
row format avro message 'user' 
row schema location 's3://schemas/complex-schema.avsc';
```

### Refer to a related PR or issue link (optional)
#5705

### CODE REVIEW

**Comment by tabVersion** [comment]
I think it is a user-facing change, please complete the corresponding part in pr description and attach a SQL that creating a source from a schema which is on s3

**Review by tabVersion — APPROVED** [review]
generally LGTM, please add complex schema related to CI

File: src/common/src/catalog/column.rs [hunk 1]
```diff
@@ -211,6 +211,38 @@ impl From<&ColumnDesc> for ProstColumnDesc {
     }
 }
 
+impl From<&DataType> for ColumnDesc {
```
Comment (BugenZhao) [inline src/common/src/catalog/column.rs:214]: I suggest removing this "general" `From` implementation as it uses dummy values for `name`, `column_id`, and other fields, which is unsuitable for other cases.

File: src/common/src/catalog/column.rs [hunk 2]
```diff
@@ -211,6 +211,38 @@ impl From<&ColumnDesc> for ProstColumnDesc {
...
+    fn from(dtype: &DataType) -> Self {
+        let fields_desc = if let DataType::Struct(fields) = dtype {
+            let mut field_descs = fields
+                .fields
+                .iter()
+                .map(ColumnDesc::from)
+                .collect::<Vec<ColumnDesc>>();
```
Comment (BugenZhao) [inline src/common/src/catalog/column.rs:221]:
```suggestion
                .collect_vec();
```

File: src/common/src/catalog/column.rs [hunk 3]
```diff
@@ -211,6 +211,38 @@ impl From<&ColumnDesc> for ProstColumnDesc {
...
+                    .zip_eq(fields.field_names.clone())
+                    .map(|(mut desc, name)| {
+                        desc.name = name;
+                        desc
+                    })
+                    .collect::<Vec<ColumnDesc>>();
+            }
+            field_descs
+        } else {
+            Vec::with_capacity(0)
```
Comment (BugenZhao) [inline src/common/src/catalog/column.rs:234]: May simply use `Vec::new()` or `vec![]`. For empty string, use `String::new()`, `Default::default()`, `"".to_owned()`, or `"".to_string()`.

Comment (waruto210) [reply src/common/src/catalog/column.rs:234]: Thanks, I mistakenly thought that Vec::new would cause memory allocation.

Comment (BugenZhao) [reply src/common/src/catalog/column.rs:234]: It's a `const fn` so don't worry about it. 🥰

File: src/common/src/catalog/column.rs (hunk 1 above)
Comment (waruto210) [reply src/common/src/catalog/column.rs:214]: use 2 functions instead.

**Pull request merged**